pallets/runtime/tests/src/transaction_ordering_test.rs

```diff
@@ -0,0 +1,1056 @@
+    )
+}
+
+fn create_investor_account(
+    tx_id: u32,
+    key: AccountKeyring,
+) -> (SecAccount, AccountCredentials, EncryptedAmount) {
+    let mut rng = StdRng::from_seed([10u8; 32]);
+    let token_name = b"ACME";
+    // Create accounts for the key holder.
+    let did = register_keyring_account(key).unwrap();
+
+    let (secret_account, account_id, public_account, init_balance) =
+        init_account(tx_id, &mut rng, token_name, key.public(), did);
+
+    let creds = AccountCredentials {
+        key: key,
+        did: did,
+        account_id: account_id,
+        public_account: public_account,
+    };
+
+    (secret_account, creds, init_balance)
+}
+
+#[test]
+fn settle_out_of_order() {
+    ExtBuilder::default()
+        .cdd_providers(vec![AccountKeyring::Eve.public()])
+        .set_max_legs_allowed(500)
+        .build()
+        .execute_with(|| {
+            // Setting:
+            //   - Alice is the token issuer, and has 10 assets in her supply.
+            //   - Bob has a normal account.
+            //   - Charlie is the mediator.
+            //   - Eve is the CDD provider.
+            let (
+                alice_creds,
+                alice_secret_account,
+                alice_init_balance,
+                charlie_creds,
+                charlie_secret_account,
+            ) = chain_set_up(10u128);
+
+            let (bob_secret_account, bob_creds, bob_init_balance) =
+                create_investor_account(2u32, AccountKeyring::Bob);
+
+            let tx_id = 1000u32;
+            // tx_id:1000 => Alice sends 5 assets to Bob.
+            // tx_id:1001 => Alice sends 3 assets to Bob.
+            //            => Charlie (the mediator) approves tx_id:1001 first.
+            //            => Charlie (the mediator) approves tx_id:1000 second.
+            let (instruction_counter1000, alice_sent_amount_1000, bob_received_amount_1000) =
+                initialize_transaction(
+                    alice_secret_account.clone(),
+                    alice_creds.clone(),
+                    alice_init_balance.clone(),
+                    bob_secret_account.clone(),
+                    bob_creds.clone(),
+                    charlie_creds.clone(),
+                    tx_id.clone(),
+                    5,
+                );
+
+            let alice_init_balance2 = alice_init_balance - alice_sent_amount_1000;
+            let tx_id = tx_id + 1;
+            let (instruction_counter1001, alice_sent_amount_1001, bob_received_amount_1001) =
+                initialize_transaction(
+                    alice_secret_account.clone(),
+                    alice_creds.clone(),
+                    alice_init_balance2.clone(),
+                    bob_secret_account.clone(),
+                    bob_creds.clone(),
+                    charlie_creds.clone(),
+                    tx_id.clone(),
+                    3,
+                );
+
+            // Approve and process tx:1001.
+            finalize_transaction(
```

Comment:
What if bob approves in order, but mediator approves out of order?

Reply:
This is exactly doing that. In `finalize_transaction()`, the mediator approves and finalizes the transaction. This is the order of authorizations:
```
// tx_id:1000 => Alice sends 5 assets to Bob. (Alice and Bob approve)
// tx_id:1001 => Alice sends 3 assets to Bob. (Alice and Bob approve)
//            => Charlie (the mediator) approves tx_id:1001 first.
//            => Charlie (the mediator) approves tx_id:1000 second.
```

Reply:
I see, the naming confused me. Since we use "finalize" to refer to receiver's action...

Reply:
Yeah, bad naming on my part.